workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt

```diff
@@ -0,0 +1,300 @@
+   * [StateRegistryAggregator]. Typically this is derived from the
+   * [compatibility key][com.squareup.workflow1.ui.Compatible.keyFor] of the [view]'s
+   * rendering.
+   */
+  public fun installChildSavedStateRegistryControllerOn(
+    view: View,
+    key: String
+  ) {
+    val lifecycleOwner = checkNotNull(WorkflowLifecycleOwner.get(view)) {
+      "Expected container view to have set a WorkflowLifecycleOwner on its immediate " +
+        "child $view."
+    }
+    val registryOwner = KeyedStateRegistryOwner(key, lifecycleOwner)
+    children.put(key, registryOwner)?.let {
+      throw IllegalArgumentException("$key is already in use, it cannot be used register $view")
+    }
+    ViewTreeSavedStateRegistryOwner.get(view)?.let {
+      "$view already has ViewTreeSavedStateRegistryOwner: $it"
+    }
+    ViewTreeSavedStateRegistryOwner.set(view, registryOwner)
+    restoreChildSavedStateRegistryOwnerIfReady(registryOwner)
+  }
+
+  /**
+   * Call this when the [View] previously [registered][installChildSavedStateRegistryControllerOn]
+   * with [key] is being dropped, but may be replaced with a new instance before the container
+   * is destroyed -- think of pushing and popping in a back stack.
+   *
+   * The saved state will be restored if a new [View] is
+   * [registered][installChildSavedStateRegistryControllerOn] with the same [key].
+   */
+  public fun saveChildSavedStateRegistryOwner(key: String) {
+    children.remove(key)?.let { saveChildSavedStateRegistryOwner(it) }
+      ?: throw IllegalArgumentException("No such child: $key")
+  }
+
+  private fun saveChildSavedStateRegistryOwner(child: KeyedStateRegistryOwner) {
+    doIfRestored { states ->
+      val state = Bundle()
+      child.controller.performSave(state)
+      states += child.key to state
+    }
+  }
+
+  /**
+   * If this object has been restored from its parent registry, restores the child
+   * [KeyedStateRegistryOwner.controller] that was previously [saved][saveChildSavedStateRegistryOwner]
+   * with the same [KeyedStateRegistryOwner.key].
+   *
+   * This method can only be called once per key per instance of this class.
+   * After a key has been restored, its data will be removed from this object.
+   */
+  private fun restoreChildSavedStateRegistryOwnerIfReady(
+    child: KeyedStateRegistryOwner
+  ) {
+    doIfRestored { states ->
+      val state = states.remove(child.key)
+      child.controller.performRestore(state)
+    }
+  }
+
+  /**
+   * Drops all child [ViewTreeSavedStateRegistryOwner]s and their restored
+   * state, except those with keys in [keysToKeep].
+   */
+  public fun prune(keysToKeep: Collection<String> = emptyList()) {
+    (children.keys - keysToKeep).forEach { children.remove(it) }
+
+    doIfRestored { states ->
+      val deadKeys = states.keys - keysToKeep
+      states -= deadKeys
+    }
+  }
+
+  private inline fun doIfRestored(block: (MutableMap<String, Bundle>) -> Unit) {
+    states?.let(block)
```

Comment:
Don't we have `isRestored` also? We are kind of duplicating that here.

Reply:
If we check `isRestored` here we'd be duplicating the `states == null` check instead.